### PR TITLE
Add reserve flag to get_volume_connector

### PR DESCRIPTION
**Pull request by changzhi1990**
In the function get_volume_connector, if the reserve flag is True,
reserve the FCP device. If the reserve flag is False and the FCP's
connections is 0 then unreserve it.

### CODE REVIEW

File: zvmsdk/volumeop.py [hunk 1]
```diff
@@ -566,6 +566,9 @@ def _rollback_dedicated_fcp(self, fcp_list, assigner_id):
                 connections = self.fcp_mgr.decrease_fcp_usage(fcp, assigner_id)
                 if connections == 0:
                     self._undedicate_fcp(fcp, assigner_id)
+                    # Unreserve the FCP device
+                    LOG.info("Unreserve fcp device %s", fcp)
+                    self.db.unreserve(fcp)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:571]: if the first fcp dedicate failed, the second fcp will still be left as reserved...

File: zvmsdk/volumeop.py [hunk 2]
```diff
@@ -754,6 +757,15 @@ def get_volume_connector(self, assigner_id):
             return empty_connector
         wwpns = []
         for fcp_no in fcp_list:
+            if reserve == True:
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:760]: no "reserve" parm in this function.

File: zvmsdk/volumeop.py [hunk 3]
```diff
@@ -754,6 +757,15 @@ def get_volume_connector(self, assigner_id):
             return empty_connector
         wwpns = []
         for fcp_no in fcp_list:
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:759]: In  line 753, the get_available_fcp function would still reserve the fcp...
You should remove that reserve.
All reserve should be done in the following logic.

File: zvmsdk/api.py [hunk 4]
```diff
@@ -1423,7 +1423,7 @@ def vswitch_delete(self, vswitch_name, persist=True):
         """
         self._networkops.delete_vswitch(vswitch_name, persist)
 
-    def get_volume_connector(self, userid):
+    def get_volume_connector(self, userid, reserve):
```
Comment (dongyanyang) [inline zvmsdk/api.py:1426]: should add a default value to this new param: reserve=False

File: zvmconnector/restclient.py [hunk 5]
```diff
@@ -430,7 +430,13 @@ def req_volume_refresh_bootmap(start_index, *args, **kwargs):
 
 def req_get_volume_connector(start_index, *args, **kwargs):
     url = '/volumes/conn/%s'
-    body = None
+    reserve = kwargs['reserve']
```
Comment (dongyanyang) [inline zvmconnector/restclient.py:433]: reserve param should be optional to keep compatibility.
If not provided, should default to False

File: zvmconnector/restclient.py [hunk 6]
```diff
@@ -430,7 +430,13 @@ def req_volume_refresh_bootmap(start_index, *args, **kwargs):
...
+    body = {'info':
```
Comment (dongyanyang) [inline zvmconnector/restclient.py:434]: I'm not sure whether it is appropriate to add this param in body as this is a GET request.

File: zvmconnector/restclient.py [hunk 7]
```diff
@@ -430,7 +430,13 @@ def req_volume_refresh_bootmap(start_index, *args, **kwargs):
...
+        {
+            "reserve": reserve
```
Comment (dongyanyang) [inline zvmconnector/restclient.py:436]: and if this is a param, we need to add some schema check

File: zvmsdk/volumeop.py [hunk 8]
```diff
@@ -754,6 +757,15 @@ def get_volume_connector(self, assigner_id):
...
+            if reserve == True:
+                # Reserve fcp device
+                LOG.info("Reserve fcp device %s", fcp_no)
+                self.db.reserve(fcp_no)
+            elif reserve == False and \
+                self.db.get_connections_from_fcp(fcp_no) == 0:
+                # Unreserve fcp device
+                LOG.info("Unreserve fcp device %s", fcp_no)
+                self.db.unreserve(fcp_no)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:768]: I don't see your change to the detach function.
Don't you need in detach function:
if connections==0:
   unreserve?

File: zvmsdk/volumeop.py [hunk 9]
```diff
@@ -558,13 +558,24 @@ def _add_disk(self, fcp, assigner_id, target_wwpns, target_lun,
                                       target_lun, multipath, os_version,
                                       mount_point, new)
 
-    def _rollback_dedicated_fcp(self, fcp_list, assigner_id):
+    def _rollback_dedicated_fcp(self, fcp_list, assigner_id,
+                                all_fcp_list=None):
+        # If attach volume fails, we need to unreserve all FCP devices.
+        if all_fcp_list:
+            for fcp in all_fcp_list:
+                if not self.db.get_connections_from_fcp(fcp):
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:566]: you should move the unreserve after the original rollback which will decrease the connections.

File: zvmsdk/api.py [hunk 10]
```diff
@@ -1440,7 +1440,7 @@ def get_volume_connector(self, userid):
 
         :param str userid: the user id of the guest
```
Comment (bjhuangr) [inline zvmsdk/api.py:1441]: add new param description in api __doc__

File: zvmconnector/restclient.py (hunk 7 above)
Comment (bjhuangr) [reply zvmconnector/restclient.py:436]: should define new param in zvmsdk/sdkwsgi/schemas

File: zvmsdk/volumeop.py [hunk 11]
```diff
@@ -558,14 +556,21 @@ def _add_disk(self, fcp, assigner_id, target_wwpns, target_lun,
...
+                                all_fcp_list=None):
         # fcp param should be a list
         for fcp in fcp_list:
             with zvmutils.ignore_errors():
                 LOG.info("Rolling back dedicated FCP: %s" % fcp)
                 connections = self.fcp_mgr.decrease_fcp_usage(fcp, assigner_id)
                 if connections == 0:
                     self._undedicate_fcp(fcp, assigner_id)
+        # If attach volume fails, we need to unreserve all FCP devices.
+        if all_fcp_list:
```
Comment (bjhuangr) [inline zvmsdk/volumeop.py:569]: any other choices to avoid full fcp list?

**Comment by changzhi1990** [comment]
please review https://github.com/openmainframeproject/python-zvm-sdk/pull/264/files.
It is based on 1.5.n

**Pull request merged**